### PR TITLE
Replace AsArgument/AsParameter with less P/Invoke-centric concepts

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Replace AsArgument/AsParameter with enum-based options to help simplify future work with non-forward-P/Invoke support (in particular unmanaged->managed support).

This change will enable us to add behavior in one place to support getting a native value from an argument value so we don't need to manually add identical (or nearly identical) behavior to every one of our marshalling generators.

It also allows us to remove some special casing for the `Forwarder` generator.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGenerator.cs [hunk 1]
```diff
@@ -20,6 +20,44 @@ public enum TargetFramework
...
+    public enum SignatureBehavior
+    {
+        /// <summary>
+        /// The native type should match the managed type, including rehydrating marshalling attributes and by-ref syntax (pure forwarding).
+        /// </summary>
+        ManagedTypeAndAttributes,
+        /// <summary>
+        /// The native signature should be the type returned by <see cref="IMarshallingGenerator.AsNativeType(TypePositionInfo)"/> passed by value.
+        /// </summary>
+        NativeType,
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGenerator.cs:35]: Not going to press on this, just asking. If this is our future public API, do we want this to be `UnmanagedType`?

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGeneratorExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.Interop
+{
+    public static class MarshallingGeneratorExtensions
+    {
+        public static TypeSyntax AsReturnType(this IMarshallingGenerator generator, TypePositionInfo info)
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGeneratorExtensions.cs:15]: Doc?

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGeneratorExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+            if (rehydratedAttributes.Count > 0)
+            {
+                param = param.AddAttributeLists(AttributeList(SeparatedList(rehydratedAttributes)));
+            }
+
+            return param;
+        }
+
+
+        private static bool TryRehydrateMarshalAsAttribute(TypePositionInfo info, out AttributeSyntax marshalAsAttribute)
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGeneratorExtensions.cs:90]: Excellent doc in method. Thank you!

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGenerator.cs (hunk 1 above)
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGenerator.cs:35]: If we change it, we should do a change through the whole `IMarshallingGenerator` API to change from `Native` to `Unmanaged`. I'm fine doing that, but I'd rather do that in a separate PR.

**Pull request merged**